### ISSUE TITLE
Only a single (grafted) commit left in branch after checkout with extension?

### ISSUE THREAD

**Issue by Simran-B** [microsoft/vscode-pull-request-github#1044]
- Extension version: 0.5.0
- VSCode Version: 1.32.1
- OS: Windows 10 64-bit

Steps to Reproduce:

Unclear. I tried to reproduce the problem with a simplified setup, but to no avail. Here is a list of things I remember:

- Had accidentally staged all changed/untracked files
- Unstaged them one by one by clicking the `-` button on them in source code management sidebar in quick succession (in parallel? it tooks quite a while until it finished in the background)
- Installed GitHub extension
- The working copy had branch `devel` checked out, potentially not up-to-date with remote
- I wanted to edit something a feature branch call `documentation/graph-traversal-build-fix`
- There was already a local branch with that name, but it was behind the remote branch (same name)
  - I had merged `origin/devel` into `documentation/graph-traversal-build-fix` on a different machine
  - My local branch was behind but that merge commit
- Via the GitHub extension, I located the corresponding PR for that branch and clicked the Checkout button
- When I checked `git log` the next time in a command prompt, it only gave me this:
  ```
  >git log
  commit b7aa332a80c16be8c74c51fb7c2f7cdb8e4e8ff2 (grafted, HEAD -> documentation/graph-traversal-build-fix, origin/documentation/graph-traversal-build-fix)
  Author: Simran ...
  Date:   Fri Mar 8 15:18:44 2019 +0100
  
      Merge remote-tracking branch 'origin/devel' into documentation/graph-traversal-build-fix

  ```
  So the entire history of that branch looks like a single commit and is the merge commit (the one I was behind before the Checkout via the GitHub extension)
- Shortly after, I noted that it said `grafted`. In a shallow clone of a repository, it would be expected that you see a truncated history. However, my repo wasn't a shallow clone but a normal working copy with the entire history.
- The PR on GitHub based on the `documentation/graph-traversal-build-fix` branch looked alright, so I deleted my local branch and checked out the remote branch - but the history remained a single commit
- Then I tried to `reset --hard`, but this didn't change anything either

I cloned the entire repo to a new folder, and there the branch is fine, i.e. the entire history is listed as expected. Does this mean some cache in the original working copy got corrupted?

What does the GitHub extension do exactly if you check out pull requests? Is there any chance for breaking the working copy as described? Is it caused by how the extension uses Git, or does this sound like a bug in Git itself?

If you want to analyze the (possibly) corrupted working copy, download it here: [arangodb-grafted.7z (1.1 GB)](https://drive.google.com/file/d/1Q7RQBhjfAsLdXoC8NOsLtgtiL6DzEnQK/view?usp=sharing)

**Comment by sunsations** [comment]
@Simran-B I had the same problem. Solution was to delete the `.git/shallow` file.

**Comment by Simran-B** [comment]
There's indeed a shallow file which contains a single line in my case:
`b7aa332a80c16be8c74c51fb7c2f7cdb8e4e8ff2` - which is also the commit it shows as entire history on the affected branch. I renamed the file and the full history is back! Thanks @sunsations for this suggestion!

A question remains: what created this shallow file or added this commit to it respectively? Is there any git command to add to this file? As I did nothing out of the ordinary with Git exception using the Github extension of VS Code, I still believe that this was caused by the extension, and I consider that to be a bug.

**Comment by Stuk** [comment]
I got the same problem when I started reviewing a PR through the extension. I have no idea how it got there.

**Comment by Stuk** [comment]
I ran into this again. I believe it's because of the fetch with a depth of 1:

https://github.com/microsoft/vscode-pull-request-github/blob/2285fdc17d20ce325163c23215f5fabc7a923be9/src/github/pullRequestGitHelper.ts#L42

This really interferes with reviewing the log of pull requests. Is it possible to make this configurable?

**Comment by MetRonnie** [comment]
This has completely messed up the `git log --graph` for my branch tracking a colleague's PR. It no longer connects between the latest commit and a previous commit on the PR. I have tried deleting `.git/shallow` and removed and re-added the remote, fetching etc. Nothing has fixed it.

**Comment by deepskyblue86** [comment]
Same here. `git fetch --unshallow` did the job.

**Comment by alexr00** [comment]
We do try to unshallow the branch now. If folks are still seeing this please ping the issue and I'll investigate!

**Comment by woutersmeenk** [comment]
@alexr00 I am still seeing this. 

- Extension version: 0.5.0
- VSCode Version: 1.77.1
- OS: MacOS 13.2.1 (22D68) (MacBook Air M1 2020)

When I checkout a branch using this extension and the local branch did not exist yet it creates the branch but only the first commit is shown with grafted tag as said above. `git fetch --unshallow` fixes this.

**Comment by alexr00** [comment]
Thanks for the heads up @woutersmeenk. Next time you see this can you share the output from GitHub Pull Request?

**Comment by woutersmeenk** [comment]
@alexr00 
I was not able to reproduce it anymore. But I found this in the logs from this morning:

```
2023-04-14 07:44:02.113 [info] Review> Switch to Pull Request #6888 - start
2023-04-14 07:44:02.168 [info] PullRequestGitHelper> Branch origin/fix/boyscouts doesn't exist on local disk yet.
2023-04-14 07:44:02.168 [info] PullRequestGitHelper> Fetch tracked branch refs/remotes/origin/fix/boyscouts
2023-04-14 07:44:04.035 [info] PullRequestGitHelper> associate fix/boyscouts with Pull Request #6888
2023-04-14 07:44:04.035 [info] Unshallowing failed: Failed to execute git {
  "exitCode": 1,
  "gitCommand": "pull",
  "stdout": "",
  "stderr": "fatal: --unshallow on a complete repository does not make sense\n"
}.
2023-04-14 07:44:04.035 [info] Falling back to git pull.
2023-04-14 07:44:04.035 [error] Pull after failed unshallow still failed: Failed to execute git {
  "exitCode": 1,
  "gitCommand": "pull",
  "stdout": "",
  "stderr": "fatal: --unshallow on a complete repository does not make sense\n"
}
2023-04-14 07:44:04.578 [info] Review> Switch to Pull Request #6888 - done
2023-04-14 07:44:04.578 [info] Review> Validate state in progress
2023-04-14 07:44:04.578 [info] Review> Validating state...
2023-04-14 07:44:04.579 [info] Found GitHub remote for folder /Users/wouter/Projects/co2m
2023-04-14 07:44:04.580 [info] Trying to use globalState for mentionable users.
2023-04-14 07:44:04.589 [info] Using globalState mentionable users for 1.
2023-04-14 07:44:04.617 [info] Review> current branch fix/boyscouts is associated with pull request #6888
2023-04-14 07:44:04.617 [info] Review> Resolving pull request
```

I think this was the first time it failed for me.

**Comment by Xottab-DUTY** [comment]
> We do try to unshallow the branch now. If folks are still seeing this please ping the issue and I'll investigate!

@alexr00, I'm still seeing this too. Any ways I can help?

VS Code 1.83.1,
Extension version 0.74.1

**Comment by alexr00** [comment]
Folks who are seeing this: are you using GitHub Codespaces?

**Comment by psychedelicious** [comment]
Hi @alexr00 , I just ran into this. I'm not using GitHub Codespaces.

I ran the command palette command "Checkout Pull Request By Number", and ended up in this state. `git fetch --unshallow` fixed it.

Here are is the output:

```sh
2024-01-28 19:20:33.315 [info] Review+0> Switch to Pull Request #5584 - start
2024-01-28 19:20:33.878 [info] PullRequestGitHelper> associate chore/update-ui-library with Pull Request #5584
2024-01-28 19:20:34.946 [info] Review+0> Switch to Pull Request #5584 - done
2024-01-28 19:20:34.946 [info] Review+0> Validate state in progress
2024-01-28 19:20:34.946 [info] Review+0> Validating state...
2024-01-28 19:20:35.007 [info] Found GitHub remote for folder /home/bat/Documents/Code/InvokeAI
2024-01-28 19:20:35.009 [info] Trying to use globalState for mentionableUsers.
2024-01-28 19:20:35.009 [info] Trying to use globalState for assignableUsers.
2024-01-28 19:20:35.040 [info] Review+0> Found matching pull request metadata for current branch chore/update-ui-library. Repo: invoke-ai/InvokeAI PR: 5584
2024-01-28 19:20:35.040 [info] Review+0> current branch chore/update-ui-library is associated with pull request #5584
2024-01-28 19:20:35.040 [info] Review+0> Resolving pull request
2024-01-28 19:20:35.040 [info] FolderRepositoryManager+0> Found GitHub repo for pr #5584: yes
2024-01-28 19:20:35.104 [info] Using globalState assignableUsers for 2.
2024-01-28 19:20:35.128 [info] Using globalState mentionableUsers for 2.
2024-01-28 19:20:36.239 [info] FolderRepositoryManager+0> Found GitHub pr repo for pr #5584: yes
2024-01-28 19:20:36.239 [info] GitHubRepository> Fetch branch main - enter
2024-01-28 19:20:36.755 [info] GitHubRepository> Fetch branch main - done: true
2024-01-28 19:20:36.755 [info] Review+0> Resolved PR #5584, state is 0
2024-01-28 19:20:36.755 [info] Review+0> Fetching pull request data
2024-01-28 19:20:36.755 [info] PullRequestTree> Adding PR #5584 to tree
2024-01-28 19:20:36.824 [info] Review+0> Register comments provider
2024-01-28 19:20:36.839 [info] Review+0> Register in memory content provider
2024-01-28 19:20:36.839 [info] FolderRepositoryManager+0> Found GitHub repo for pr #5584: yes
2024-01-28 19:20:37.017 [info] PullRequestTree> Creating file and commit nodes for PR #5584
2024-01-28 19:20:37.030 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:20:37.030 [info] PullRequestTree> Got all children for Files node
2024-01-28 19:20:37.346 [info] PullRequestTree> Review threads have changed, refreshing Files node
2024-01-28 19:20:37.346 [info] PullRequestTree> Review threads have changed, refreshing Commits node
2024-01-28 19:20:37.366 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:20:37.366 [info] PullRequestTree> Got all children for Files node
2024-01-28 19:20:37.552 [info] FolderRepositoryManager+0> Found GitHub pr repo for pr #5584: yes
2024-01-28 19:20:37.552 [info] GitHubRepository> Fetch branch main - enter
2024-01-28 19:20:37.990 [info] GitHubRepository> Fetch branch main - done: true
2024-01-28 19:20:43.005 [info] PullRequestTree> Local files have changed, refreshing Files node
2024-01-28 19:20:43.005 [info] PullRequestTree> Local files have changed, refreshing Files node
2024-01-28 19:20:43.130 [info] Review+0> Display pull request status bar indicator.
2024-01-28 19:20:43.130 [info] Review+0> Using focus mode = true.
2024-01-28 19:20:43.130 [info] Review+0> State validation silent = false.
2024-01-28 19:20:43.130 [info] Review+0> PR show should show = false.
2024-01-28 19:20:43.272 [info] GitContentFileSystemProvider> Getting change model (file:///home/bat/Documents/Code/InvokeAI) content for commit de20711637b7e04e7e1ce16cf06d41002542d897 and path /home/bat/Documents/Code/InvokeAI/invokeai/frontend/web/.eslintrc.js
2024-01-28 19:20:43.325 [info] GitContentFileSystemProvider> Getting repository (file:///home/bat/Documents/Code/InvokeAI) content for commit de20711637b7e04e7e1ce16cf06d41002542d897 and path /home/bat/Documents/Code/InvokeAI/invokeai/frontend/web/.eslintrc.js
2024-01-28 19:20:43.360 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:20:43.360 [info] PullRequestTree> Got all children for Files node
2024-01-28 19:20:43.783 [info] PullRequestTree> Comments have changed, refreshing Files node
2024-01-28 19:20:43.783 [info] PullRequestTree> Comments have changed, refreshing Commits node
2024-01-28 19:20:43.783 [info] PullRequestTree> Review threads have changed, refreshing Files node
2024-01-28 19:20:43.783 [info] PullRequestTree> Review threads have changed, refreshing Commits node
2024-01-28 19:20:44.576 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:20:44.576 [info] PullRequestTree> Got all children for Files node
2024-01-28 19:20:45.135 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:20:45.136 [info] PullRequestTree> Got all children for Files node
# --------------------Ran `git fetch --unshallow` here------------------
2024-01-28 19:24:47.219 [info] FolderRepositoryManager+0> Found GitHub repo for pr #5584: yes
2024-01-28 19:24:48.152 [info] FolderRepositoryManager+0> Found GitHub pr repo for pr #5584: yes
2024-01-28 19:24:48.152 [info] GitHubRepository> Fetch branch main - enter
2024-01-28 19:24:48.653 [info] GitHubRepository> Fetch branch main - done: true
2024-01-28 19:24:54.108 [info] PullRequestTree> Local files have changed, refreshing Files node
2024-01-28 19:24:54.108 [info] PullRequestTree> Local files have changed, refreshing Files node
2024-01-28 19:24:54.398 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:24:54.399 [info] PullRequestTree> Got all children for Files node
2024-01-28 19:24:54.728 [info] PullRequestTree> Comments have changed, refreshing Files node
2024-01-28 19:24:54.728 [info] PullRequestTree> Comments have changed, refreshing Commits node
2024-01-28 19:24:54.728 [info] PullRequestTree> Review threads have changed, refreshing Files node
2024-01-28 19:24:54.728 [info] PullRequestTree> Review threads have changed, refreshing Commits node
2024-01-28 19:24:54.886 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:24:54.886 [info] PullRequestTree> Got all children for Files node
2024-01-28 19:24:55.179 [info] PullRequestTree> Getting children for Files node
2024-01-28 19:24:55.179 [info] PullRequestTree> Got all children for Files node
```

VSCode Version:

```
Version: 1.85.2
Commit: 8b3775030ed1a69b13e4f4c628c612102e30a681
Date: 2024-01-18T06:40:11.430Z
Electron: 25.9.7
ElectronBuildId: 26354273
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Darwin arm64 23.0.0
```

GH Pull Requests extension version:
```
v0.78.1
```

Hope this helps.

**Comment by alexr00** [comment]
@psychedelicious thanks for the heads up! I just removed the shallow fetch that we do last week, so this shouldn't be a problem after the new stable release later this week.
